browser/models.py

```diff
@@ -46,56 +74,92 @@ class Project(db.Model):
     createdAt = db.Column(db.TIMESTAMP, nullable=False, default=db.func.now())
     finished = db.Column(db.TIMESTAMP)
 
+    filename = db.Column(db.Text, nullable=False)
+    path = db.Column(db.Text, nullable=False)
+    output_bucket = db.Column(db.Text, nullable=False)
+    height = db.Column(db.Integer, nullable=False)
+    width = db.Column(db.Integer, nullable=False)
+    num_frames = db.Column(db.Integer, nullable=False)
+    num_channels = db.Column(db.Integer, nullable=False)
+    num_features = db.Column(db.Integer, nullable=False)
+    rgb = db.Column(db.Boolean, default=False)
+    frame = db.Column(db.Integer, default=0)
+    channel = db.Column(db.Integer, default=0)
+    feature = db.Column(db.Integer, default=0)
+    scale_factor = db.Column(db.Float, default=1)
+    colormap = db.Column(db.PickleType)
+
     raw_frames = db.relationship('RawFrame', backref='project')
     rgb_frames = db.relationship('RGBFrame', backref='project')
-    label_frames = db.relationship('LabelFrame', backref='project')
-    state = db.relationship('State', backref='project', uselist=False)
+    label_frames = db.relationship('LabelFrame', backref='project',
+                                   # Delete frames detached by undo/redo
+                                   cascade='save-update, merge, delete, delete-orphan')
+    labels = db.relationship('Labels', backref='project', uselist=False,
+                             # Delete labels detached by undo/redo
+                             cascade='save-update, merge, delete, delete-orphan')
+
+    # Action history
+    action_id = db.Column(db.Integer)
+    next_action_id = db.Column(db.Integer, default=0)
+    actions = db.relationship('Action', backref='project')
 
     def __init__(self, filename, input_bucket, output_bucket, path,
-                 rgb=False, raw_key='raw', annotated_key=None):
+                 raw_key='raw', annotated_key=None):
         init_start = timeit.default_timer()
+
+        # Load data
         if annotated_key is None:
             annotated_key = get_ann_key(filename)
-
         start = timeit.default_timer()
         trial = self.load(filename, input_bucket, path)
-        current_app.logger.debug('Loaded file %s from S3 in %s s.',
-                                 filename, timeit.default_timer() - start)
+        logger.debug('Loaded file %s from S3 in %ss.',
+                     filename, timeit.default_timer() - start)
```

Comment:
Why not `current_app`?

Reply:
Previously when models.py just contained one table that stored just the pickled "state" (ZStackEdit/TrackEdit object), the only log statement in models.py used the logger defined in line 28 in models.py `logger = logging.getLogger('models.Project')`.

As I started moving things from caliban.py into models.py, the current_app logger moved with them, but just before submitting the PR, I noticed the two loggers and went with the original one for consistency.

I don't quite understand the difference between these two loggers and I'd be happy to change all the log statements back to `current_app.logger`, as long as we're consistent in which we're using.

Reply:
Ah that makes a lot of sense. I agree, consistency is the most important. Let's keep it as the model/module level logger in that case. Thanks for clearing that up!